Client/Frontend/Browser/BrowserToolbar.swift

```diff
@@ -4,6 +4,170 @@
 
 import Foundation
 import UIKit
+import Snappy
+
+protocol BrowserLocationViewDelegate {
+    func browserLocationViewDidTapLocation(browserLocationView: BrowserLocationView)
+    func browserLocationViewDidTapReaderMode(browserLocationView: BrowserLocationView)
+}
+
+class ReaderModeButton: UIButton {
+    override init(frame: CGRect) {
+        super.init(frame: frame)
+        setImage(UIImage(named: "reader.png"), forState: UIControlState.Normal)
+        setImage(UIImage(named: "reader_active.png"), forState: UIControlState.Selected)
+    }
+
+    required init(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    var _readerModeState: ReaderModeState = ReaderModeState.Unavailable
+
+    var readerModeState: ReaderModeState {
+        get {
+            return _readerModeState;
+        }
+        set (newReaderModeState) {
+            _readerModeState = newReaderModeState
+            switch _readerModeState {
+            case .Available:
+                self.enabled = true
+                self.selected = false
+            case .Unavailable:
+                self.enabled = false
+                self.selected = false
+            case .Active:
+                self.enabled = true
+                self.selected = true
+            }
+        }
+    }
+}
+
+class BrowserLocationView : UIView, UIGestureRecognizerDelegate {
```

Comment:
Filed bugs for moving them out of this file.I think it will be good to have them available non-private because that means we can use them in future unit tests for example.